src/main/software/amazon/event/ruler/ComparableNumber.java

```diff
@@ -3,104 +3,122 @@
 import ch.randelshofer.fastdoubleparser.JavaBigDecimalParser;
 
 import java.math.BigDecimal;
-
-import static software.amazon.event.ruler.Constants.BASE64_DIGITS;
-import static software.amazon.event.ruler.Constants.MIN_NUM_DIGIT;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.List;
 
 /**
  * Represents a number as a comparable string.
  * <br/>
- * Numbers are allowed in the range -500,000,000,000 to +500,000,000,000 (inclusive).
- * Comparisons are precise to 17 decimal places, with six to the right of the decimal.
- * Numbers are treated as floating-point values.
- * <br>
- * Numbers are converted to strings by:
- * 1. Multiplying by 1,000,000 to remove the decimal point and then adding 500,000,000,000 (to remove negatives), then
- * 2. Formatting to a 12-character to base64 string with padding, because the base64 string
- *     converted from 500,000,000,000 * 1,000,000 = 500,000,000,000,000,000 has 12 characters.
+ * All possible double numbers (IEEE-754 binary64 standard) are allowed.
+ * Numbers are first standardized to floating-point values and then converted
+ * to a Base128 encoded string of 10 bytes.
  * <br/>
- * Hexadecimal representation is used because:
- * 1. It saves 3 bytes of memory per number compared to decimal representation.
- * 2. It is lexicographically comparable, which is useful for maintaining sorted order of numbers.
- * 2. It aligns with the radix used for IP addresses.
+ * We use Base128 encoding offers a compact representation of decimal numbers
+ * as it preserves the lexicographical order of the numbers. See
+ * https://github.com/aws/event-ruler/issues/179 for more context.
  * <br/>
- * The number is parsed as a Java {@code BigDecimal} to support decimal fractions. We're avoiding double as
- * there is a well-known issue that double numbers can lose precision when performing calculations involving
- * other data types. The higher the number, the lower the accuracy that can be maintained. For example,
- * {@code 0.30d - 0.10d = 0.19999999999999998} instead of {@code 0.2d}. When extended to {@code 1e10}, the test
- * results show that only 5 decimal places of precision can be guaranteed when using doubles.
+ * The numbers are first parsed as a Java {@code BigDecimal} as there is a well known issue
```

Comment:
Can you provide a reference or details for that?

Double handling is finicky and has lots of potential foot guns. But I never ran into a loss of precision that's not mediated by decimal-binary conversion issues. If you e.g. use Double.toHexString and the "p" representation of exponents in your numerical tests, you should be able to avoid these issues.
This only applies to where you want to use numbers and not the strings, of course. But at some point, the JSON number strings will be converted to double and you'll run into the same issue.
Unless the JSON parsers also avoid Double. But then, you should use the same code path the JSON parser uses.

You won't lose fidelity. But you'll gain some performance.

Reply:
`37.807807921694092` was the number I noticed it with.

Reply:
A double represented as a decimal stops being accurate after 15 digits (unless it ends in 5). this number has 17.
A good way to keep it in mind: 10 bits fit 3 decimal digits. a double has 52 mantissa bits, so this buys you a fractional amount more than 15 digits with 50 bits. the other 2 (or 3 if you include the implicit one) are not enough for a 16th digit. So it will... distribute... the rest of the 10 alternatives over 8 states. As powers of two are binary, a number ending in 5 will be at an advantage, here.

Reply:
And BTW this is a problem with the citylots file, those floats in there have more precision than is safe with JSON. So it's probably not useful for supporting numeric-match testing.

Reply:
Oh wow. I did not realize that. Let me look into changing city lots to have a healthy mix of numbers with less than 15 decimals.